### PR TITLE
Fix races and skip NoPia test on CoreClr

### LINKED ISSUE #18985: Sometimes the type identity seems to fail on CoreCLR
We see this only in tests and very rarely, it is likely to be an artifact of assembly loading races in tests or something like that.

We need to investigate this and fix. Even if rare, it happens often enough to cause failed builds and other grief.

A typical failure would be a random test failing with something like:

Assert.Same() Failure 
Expected: Pia1, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null 
Actual: Pia1, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null

or 

12:17:10         Assert.Equal() Failure
12:17:10         Expected: System.IComparable
12:17:10         Actual:   System.IComparable

I.E. Seemingly same types compare as not equal where they are expected to be equal.

### PR DESCRIPTION
**Pull request by agocke**
Fixes #18985

### CODE REVIEW

**Comment by agocke** [comment]
cc @dotnet/roslyn-compiler for review

**Comment by agocke** [comment]
test windows_debug_vs-integration_prtest please

File: src/Test/Utilities/Portable/TestBase.cs [hunk 1]
```diff
@@ -237,7 +237,10 @@ public static MetadataReference MscorlibRef
             {
                 if (s_mscorlibRef == null)
                 {
-                    s_mscorlibRef = AssemblyMetadata.CreateFromImage(TestResources.NetFX.v4_0_30319.mscorlib).GetReference(display: "mscorlib.v4_0_30319.dll");
+                    Interlocked.CompareExchange(
+                        ref s_mscorlibRef,
+                        AssemblyMetadata.CreateFromImage(TestResources.NetFX.v4_0_30319.mscorlib).GetReference(display: "mscorlib.v4_0_30319.dll"),
+                        null);
```
Comment (cston) [inline src/Test/Utilities/Portable/TestBase.cs]: Is the same change needed for other `MetadataReference` properties in this file?

If so, perhaps we should have a helper: `GetMetadataReference(ref MetadataReference reference, string displayName, Func<MetadataReference> getMetadata)`.

Comment (agocke) [inline src/Test/Utilities/Portable/TestBase.cs]: As far as I can tell, only mscorlib is commonly depended on for symbol equality, which is the only meaningful difference in testing. I think we should be safe just using cmpxchg for mscorlib.

Comment (cston) [inline src/Test/Utilities/Portable/TestBase.cs]: Looks like `SystemCoreRef` uses `CompareExchange` as well.

Besides reference equality, a helper would also simplify the property getters.
```
public static MetadataReference SystemDataRef => 
    GetReference(ref s_systemDataRef, AssemblyMetadata.CreateFromImage(TestResources.NetFX.v4_0_30319.System_Data).GetReference(display: "System.Data.v4_0_30319.dll"));

Comment (agocke) [inline src/Test/Utilities/Portable/TestBase.cs]: Sure, I'll add a helper.

Comment (agocke) [inline src/Test/Utilities/Portable/TestBase.cs]: @cston Sorry for the delay, I forgot this PR was still open -- but it's ready for re-review when you get a chance.

**Comment by cston** [comment]
LGTM

**Pull request merged**